signer/src/schnorr.rs

```diff
@@ -0,0 +1,156 @@
+        S: Signable,
+    {
+        let k: ScalarField = self.blinding_hash(&kp, input);
+        let r: CurvePoint = CurvePoint::prime_subgroup_generator().mul(k).into_affine();
+        let k: ScalarField = if r.y.into_repr().is_even() { k } else { -k };
+
+        let e: ScalarField = self.message_hash(&kp.public, r.x, input);
+        let s: ScalarField = k + e * kp.secret.to_scalar();
+
+        Signature::new(r.x, s)
+    }
+
+    fn verify<S>(&mut self, sig: Signature, public: PubKey, input: S) -> bool
+    where
+        S: Signable,
+    {
+        let ev: ScalarField = self.message_hash(&public, sig.rx, input);
+
+        let sv: CurvePoint = CurvePoint::prime_subgroup_generator()
+            .mul(sig.s)
+            .into_affine();
+        // Perform addition and infinity check in projective coordinates for performance
+        let rv = public.to_point().mul(ev).neg().add_mixed(&sv);
+        if rv.is_zero() {
+            return false;
+        }
+        let rv = rv.into_affine();
+
+        rv.y.into_repr().is_even() && rv.x == sig.rx
+    }
+}
+
+impl<SC: SpongeConstants> Schnorr<SC> {
+    /// Create a new Schnorr signer context for network instance `network_id` using arithmetic sponge defined by `sponge`.
+    pub fn new(sponge: ArithmeticSponge<BaseField, SC>, network_id: NetworkId) -> Schnorr<SC> {
+        Schnorr::<SC> { sponge, network_id }
+    }
+
+    fn domain_bytes<S>(network_id: NetworkId) -> Vec<u8>
+    where
+        S: Signable,
+    {
+        let mut domain_string = S::domain_string(network_id);
+        // Domain prefixes have a max length of 20 and are padded with '*'
+        assert!(domain_string.len() <= 20);
+        domain_string = &domain_string[..std::cmp::min(domain_string.len(), 20)];
+        let mut bytes = format!("{:*<20}", domain_string).as_bytes().to_vec();
+        bytes.resize(32, 0);
+
+        bytes
+    }
+
+    // This function uses a cryptographic hash function to create a uniformly and
+    // randomly distributed nonce.  It is crucial for security that no two different
+    // messages share the same nonce.
+    fn blinding_hash<H>(&self, kp: &Keypair, input: H) -> ScalarField
+    where
+        H: Hashable,
+    {
+        let mut hasher = VarBlake2b::new(32).unwrap();
+
+        let mut roi: ROInput = input.to_roinput();
+        roi.append_field(kp.public.to_point().x);
+        roi.append_field(kp.public.to_point().y);
+        roi.append_scalar(kp.secret.to_scalar());
+        roi.append_bytes(&[self.network_id.into()]);
+
+        hasher.update(roi.to_bytes());
+
+        let mut bytes = [0; 32];
+        hasher.finalize_variable(|out| bytes.copy_from_slice(out));
+        // Drop the top two bits to convert into a scalar field element
+        //   N.B. Since the order p is very close to 2^m for some m, truncating only creates
+        //   a tiny amount of bias that should be insignificant and keeps the implementation
+        //   simple by avoiding reduction modulo p.
```

Comment:
two other things you can point out:

- that this is true for Pallas' scalar field at least (not necessarily true if we change curve)
- that doing a modular reduction would be worse

Reply:
👍🏼